jsonpb/jsonpb.go

```diff
@@ -116,13 +127,43 @@ type wkt interface {
 	XXX_WellKnownType() string
 }
 
+// customJSON can be implemented by messages that have custom non-object JSON
+// representations. If the message implements this method and the method
+// returns true, the JSON representation will not be an object (for example,
+// the message marshals itself as a string).
+type customJSON interface {
+	XXX_CustomJSON() bool
+}
+
 // marshalObject writes a struct to the Writer.
 func (m *Marshaler) marshalObject(out *errWriter, v proto.Message, indent, typeURL string) error {
 	if jsm, ok := v.(JSONPBMarshaler); ok {
 		b, err := jsm.MarshalJSONPB(m)
 		if err != nil {
 			return err
 		}
+		if typeURL != "" {
+			// we are marshaling this object to an Any type
+			var js interface{}
+			if err = json.Unmarshal(b, &js); err != nil {
+				return fmt.Errorf("Type %v produced invalid JSON: %v", reflect.TypeOf(v), err)
+			}
+			m, ok := js.(map[string]interface{});
+			if ok {
+				m["@type"] = typeURL
+			} else {
+				if cj, ok := v.(customJSON); !ok || !cj.XXX_CustomJSON() {
```

Comment:
> As mentioned above, I'm not sure if this is necessary. I'd suggest error'ing here for now. 

Okay, I'll tweak this branch to remove the `customJson` stuff and push changes.

> But if you think this is something needed, then I think you can simply take the RawMessage and set it to "value", right?

I think that's what the code is doing below. But perhaps I'm not following this remark. (Though it will be moot after removing the `customJson` stuff.)